subcommands/keys/tuf_updates_add_offline_key.go

```diff
@@ -0,0 +1,124 @@
+  fioctl keys tuf updates add-offline-key \
+	 --txid=abc --role=root --keys=tuf-root-keys.tgz
+- Add offline TUF targets key, explicitly specifying new key type (and signing algorithm):
+  fioctl keys tuf updates add-offline-key \
+    --txid=abc --role=targets --keys=tuf-targets-keys.tgz --key-type=ed25519`,
+		Run: doTufUpdatesAddOfflineKey,
+	}
+	add.Flags().StringP("role", "r", "", "TUF role name, supported: Root, Targets.")
+	_ = add.MarkFlagRequired("role")
+	add.Flags().StringP("txid", "x", "", "TUF root updates transaction ID.")
+	add.Flags().StringP("keys", "k", "", `Path to <tuf-keys.tgz> where a new key should be created.
+For security reasons, it is disallowed to add several actual keys for the same TUF role into the same file.`)
+	_ = add.MarkFlagFilename("keys")
+	_ = add.MarkFlagRequired("keys")
+	add.Flags().StringP("key-type", "y", tufKeyTypeNameEd25519, "Key type, supported: Ed25519, RSA.")
+	tufUpdatesCmd.AddCommand(add)
+}
+
+func doTufUpdatesAddOfflineKey(cmd *cobra.Command, args []string) {
+	factory := viper.GetString("factory")
+	txid, _ := cmd.Flags().GetString("txid")
+	keyTypeStr, _ := cmd.Flags().GetString("key-type")
+	keyType := ParseTufKeyType(keyTypeStr)
+	keysFile, _ := cmd.Flags().GetString("keys")
+	roleName, _ := cmd.Flags().GetString("role")
+	roleName = ParseTufRoleNameOffline(roleName)
+
+	var creds OfflineCreds
+	if _, err := os.Stat(keysFile); err == nil {
+		creds, err = GetOfflineCreds(keysFile)
+		subcommands.DieNotNil(err)
+		subcommands.AssertWritable(keysFile)
+	} else if errors.Is(err, fs.ErrNotExist) {
+		creds = make(OfflineCreds, 0)
+		saveTufCreds(keysFile, creds)
+	} else {
+		subcommands.DieNotNil(err)
+	}
+
+	updates, err := api.TufRootUpdatesGet(factory)
+	subcommands.DieNotNil(err)
+
+	_, newCiRoot := checkTufRootUpdatesStatus(updates, true)
+
+	switch roleName {
+	case tufRoleNameRoot:
+		addOfflineRootKey(newCiRoot, creds, keyType)
+	case tufRoleNameTargets:
+		onlineTargetsId := updates.Updated.OnlineKeys["targets"]
+		if onlineTargetsId == "" {
+			subcommands.DieNotNil(errors.New("Unable to find online target key for factory"))
+		}
+		addOfflineTargetsKey(newCiRoot, creds, keyType, onlineTargetsId)
+	default:
+		panic(fmt.Errorf("Unexpected role name: %s", roleName))
+	}
+
+	newCiRoot.Signatures = make([]tuf.Signature, 0)
+	removeUnusedTufKeys(newCiRoot)
+	newProdRoot := genProdTufRoot(newCiRoot)
+
+	fmt.Println("= Uploading new TUF root")
+	tmpFile := saveTempTufCreds(keysFile, creds)
+	err = api.TufRootUpdatesPut(factory, txid, newCiRoot, newProdRoot, nil)
+	handleTufRootUpdatesUpload(tmpFile, keysFile, err)
+}
+
+func addOfflineRootKey(root *client.AtsTufRoot, creds OfflineCreds, keyType TufKeyType) {
+	subcommands.DieNotNil(checkNoTufSigner(root, creds, root.Signed.Roles["root"].KeyIDs))
+
+	kp := genTufKeyPair(keyType)
+	root.Signed.Keys[kp.signer.Id] = kp.atsPub
+	root.Signed.Expires = time.Now().AddDate(1, 0, 0).UTC().Round(time.Second) // 1 year validity
```

Comment:
Why the root meta is refreshed (expiration reset) if a root key is being added and we don't do it if a targets key is being added?

Reply:
Not sure I get it. Adding a key doesn't mean key rotation. So, why we refresh root meta each time new root key is added to it is not clear to me.

Reply:
I thought it is not clear to you why targets key addition/rotation does not refresh expiration.
So, I answered that part of your question.

For a question why do we refresh expiration when a user adds a new key I have no clear answer.
The only argument is: because we do the same thing for the TUF root rotation, so we need a parity here.
If you think that _this is bad_ - I'm fine to remove this line, so that we only increase expires field at TUF root rotation.

In the future, I think we need to come out with a solution when _"root.json expires field equals to the minimum value of expiration of all its keys"_. But, TUF spec does not have a "standard" solution for that; so we need to invent our own wheel. That part is our of scope.

Reply:
I don't know the answer myself. On the one side, it makes sense to refresh root meta expiration each time its updated (new root meta version), then if a new root key or targets key or any other role keys is added to room meta, then its expiration should be refreshed.

On the other hand, a user may wanna have a full control over root meta expiration, then we should not refresh at all.

So, I am fine with either refresh if anything changes in root meta or don't refresh on changes unless a user specifically says so. I am not sure I get the mix approach, if a root key is added we refresh if some other role key is added we don't.

Maybe, we should just add a flag `--refresh-root` to that command.

Reply:
I guess it's not critical atm.

Reply:
After some thought I think this is the correct fixup at our current stage: https://github.com/foundriesio/fioctl/pull/241/commits/c13c0c6142c7c4b1e7d1a9262163b42c93548e90